### ISSUE TITLE
Endless search loop when using complex _filter parameter

### ISSUE THREAD

**Issue by ttntrifork** [jamesagnew/hapi-fhir#1577]
Search requests fail when using a sufficiently complex `_filter` parameter.

The following request terminates successfully and returns data:
> [http://hapi.fhir.org/baseDstu3/Appointment?_filter=actor eq Patient/2616593 or actor eq Patient/2616593](http://hapi.fhir.org/baseDstu3/Appointment?_filter=actor%20eq%20Patient/2616593%20or%20actor%20eq%20Patient/2616593)

If one more condition is added to the _filter parameter, the request is aborted after a short while:
> [http://hapi.fhir.org/baseDstu3/Appointment?_filter=actor eq Patient/2616593 or actor eq Patient/2616593 or actor eq Patient/2616593](http://hapi.fhir.org/baseDstu3/Appointment?_filter=actor%20eq%20Patient/2616593%20or%20actor%20eq%20Patient/2616593%20or%20actor%20eq%20Patient/2616593)

The same behaviour is observed on my local HAPI FHIR fork (4.1.0-SNAPSHOT), except that the timeout is after 1 minute and happens at [SearchCoordinatorSvcImpl.java#L228](https://github.com/jamesagnew/hapi-fhir/blob/cc108be93936d9a60f337b5387d892b333ce2d32/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java#L228)

I believe that the problem occurs when a search is not marked as FINISHED at [SearchCoordinatorSvcImpl.java#L791](https://github.com/jamesagnew/hapi-fhir/blob/cc108be93936d9a60f337b5387d892b333ce2d32/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java#L791), but is instead marked as PASSCMPLET which causes it to get restarted at [SearchCoordinatorSvcImpl.java#L233](https://github.com/jamesagnew/hapi-fhir/blob/cc108be93936d9a60f337b5387d892b333ce2d32/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java#L233). 
This repeats in an endless loop until the search times out.

The reason that the search is not marked as FINISHED in this case is due to the many `left outer join` in the generated SQL query which causes the same resource to be found many times in the result set, which increases the [skippedCount](https://github.com/jamesagnew/hapi-fhir/blob/237a3049577690e5afb4d792c613747c4240aa6c/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/SearchBuilder.java#L3204) to above the current [search prefetch threshold](https://github.com/jamesagnew/hapi-fhir/blob/5f17fb30b4ae62c59af8cb6068b233132bc7fe36/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/DaoConfig.java#L1345). 

For some reason the search is restarted with the exact same parameters that results in the exact same outcome, and thus loops indefinately.

**Environment (please complete the following information):**
 - HAPI FHIR Version: 4.1.0-SNAPSHOT (both on http://hapi.fhir.org and a 1 month old local snapshot)
 - OS: [e.g. iOS]: Windows
 - Browser [e.g. chrome, safari]: Chrome

**Additional context**
Example of SQL query with many left outer joins:

```sql
select resourceta0_.RES_ID as col_0_0_
from HFJ_RESOURCE resourceta0_
         left outer join HFJ_RES_LINK myresource1_ on resourceta0_.RES_ID = myresource1_.SRC_RESOURCE_ID
         left outer join HFJ_RES_LINK myresource2_ on resourceta0_.RES_ID = myresource2_.SRC_RESOURCE_ID
         left outer join HFJ_RES_LINK myresource3_ on resourceta0_.RES_ID = myresource3_.SRC_RESOURCE_ID
         left outer join HFJ_RES_LINK myresource4_ on resourceta0_.RES_ID = myresource4_.SRC_RESOURCE_ID
where (myresource1_.SRC_PATH in (?)) and (myresource1_.TARGET_RESOURCE_ID in (?))
   or (myresource2_.SRC_PATH in (?)) and (myresource2_.TARGET_RESOURCE_ID in (?))
   or (myresource3_.SRC_PATH in (?)) and (myresource3_.TARGET_RESOURCE_ID in (?))
   or (myresource4_.SRC_PATH in (?)) and (myresource4_.TARGET_RESOURCE_ID in (?))

```

**Comment by jvitrifork** [comment]
@jamesagnew If you could find time to have a look at this we would be very grateful

**Comment by jvitrifork** [comment]
@ttntrifork this seems fixed with release 5.0.0

**Comment by ttntrifork** [comment]
Yes, can no longer reproduce
